### PR TITLE
UI: camera view widget

### LINKED ISSUE #20861: UI: move driver view to Qt
First step is to make a generic camera view widget, then use that to make a nice driver view widget.

@deanlee interested in this one?

### PR DESCRIPTION
**Pull request by deanlee**

### CODE REVIEW

**Comment by adeebshihadeh** [comment]
Let's hold off on using this for the road camera for now and just do driver view in this PR. It'll be easier to review and test.

**Comment by deanlee** [comment]
got it

File: selfdrive/ui/qt/home.cc [hunk 1]
```diff
@@ -187,3 +202,98 @@ void OffroadHome::refresh() {
   }
   alert_notification->setStyleSheet(style);
 }
+
+// class DriverViewWindow
+
+DriverViewWindow::DriverViewWindow(QWidget* parent) : sm({"driverState"}), QOpenGLWidget(parent) {
```
Comment (adeebshihadeh) [inline selfdrive/ui/qt/home.cc:208]: let's move this to its own file in selfdrive/ui/qt/offroad

File: selfdrive/ui/qt/home.cc [hunk 2]
```diff
@@ -49,11 +50,25 @@ void HomeWindow::offroadTransition(bool offroad) {
   emit offroadTransitionSignal(offroad);
 }
 
+void HomeWindow::driverView() {
+  if (!driver_view) {
+    Params().putBool("IsDriverViewEnabled", true);
```
Comment (adeebshihadeh) [inline selfdrive/ui/qt/home.cc:55]: do all the params handling in the driver view class

File: selfdrive/ui/qt/home.cc [hunk 3]
```diff
@@ -187,3 +202,98 @@ void OffroadHome::refresh() {
...
+}
+
+void DriverViewWindow::initializeGL() {
+  initializeOpenGLFunctions();
+  vision = std::make_unique<UIVision>(VISION_STREAM_RGB_FRONT);
+  timer->start(0);
+}
+
+void DriverViewWindow::paintGL() {
+  const NVGcolor color = bg_colors[STATUS_DISENGAGED];
```
Comment (adeebshihadeh) [inline selfdrive/ui/qt/home.cc:233]: maybe time to make bg_colors use QColor?

File: selfdrive/ui/qt/home.cc (hunk 2 above)
Comment (adeebshihadeh) [reply selfdrive/ui/qt/home.cc:55]: on the showEvent write the param, delete on hideEvent

**Comment by adeebshihadeh** [comment]
No need to make a UIVision class, just put it all in the widget. Don't worry about duplicating code, since we'll move the road camera to use this after the PR. I'm thinking the steps are:
* merge this for driver view
* I'll finish up #20825
* move road cam to Qt

File: selfdrive/ui/qt/home.cc [hunk 4]
```diff
@@ -50,11 +50,22 @@ void HomeWindow::offroadTransition(bool offroad) {
   emit offroadTransitionSignal(offroad);
 }
 
+void HomeWindow::driverView() {
+  if (!driver_view) {
```
Comment (adeebshihadeh) [inline selfdrive/ui/qt/home.cc:54]: think it makes more sense for driver view to live in the settings window or maybe main window

File: selfdrive/ui/qt/offroad/driverview.h [hunk 5]
```diff
@@ -0,0 +1,20 @@
+#pragma once
+
+#include <memory>
+
+#include "selfdrive/ui/qt/widgets/cameraview.h"
+class DriverViewWindow : public CameraViewWidget {
```
Comment (adeebshihadeh) [inline selfdrive/ui/qt/offroad/driverview.h:6]: whitespace

File: selfdrive/ui/qt/offroad/driverview.cc [hunk 6]
```diff
@@ -0,0 +1,79 @@
+#include "selfdrive/ui/qt/offroad/driverview.h"
+
+#include <QPainter>
+
+#include "selfdrive/ui/qt/qt_window.h"
+#include "selfdrive/ui/qt/util.h"
+
+DriverViewWindow::DriverViewWindow(QWidget* parent) : sm({"driverState"}), CameraViewWidget(VISION_STREAM_RGB_FRONT, parent) {
+  is_rhd = Params().getBool("IsRHD");
```
Comment (adeebshihadeh) [inline selfdrive/ui/qt/offroad/driverview.cc:9]: read this on the showEvent

File: selfdrive/ui/qt/widgets/cameraview.cc [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+      glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_SWIZZLE_R, GL_BLUE);
+      glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_SWIZZLE_G, GL_GREEN);
+      glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_SWIZZLE_B, GL_RED);
+      assert(glGetError() == GL_NO_ERROR);
+    }
+    last_frame = nullptr;
+  }
+
+  if (vipc_client->connected) {
+    if (VisionBuf *buf = vipc_client->recv(); buf != nullptr) {
```
Comment (adeebshihadeh) [inline selfdrive/ui/qt/widgets/cameraview.cc:221]:
```suggestion
    VisionBuf *buf = vipc_client->recv();
    if (buf) {
```

File: selfdrive/ui/qt/widgets/cameraview.cc [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+      last_frame = buf;
+    } else if (!Hardware::PC()) {
```
Comment (adeebshihadeh) [inline selfdrive/ui/qt/widgets/cameraview.cc:223]: else if can be removed

File: selfdrive/ui/qt/offroad/driverview.cc [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+#include "selfdrive/ui/qt/qt_window.h"
+#include "selfdrive/ui/qt/util.h"
+
+DriverViewWindow::DriverViewWindow(QWidget* parent) : sm({"driverState"}), CameraViewWidget(VISION_STREAM_RGB_FRONT, parent) {
+  face_img = QImage("../assets/img_driver_face").scaled(180, 180, Qt::KeepAspectRatio, Qt::SmoothTransformation);
+}
+
+void DriverViewWindow::showEvent(QShowEvent* event) {
+  is_rhd = Params().getBool("IsRHD");
+  Params().putBool("IsDriverViewEnabled", true);
```
Comment (adeebshihadeh) [inline selfdrive/ui/qt/offroad/driverview.cc:14]: use the same params instance everwhere

File: selfdrive/ui/qt/home.cc (hunk 4 above)
Comment (adeebshihadeh) [reply selfdrive/ui/qt/home.cc:54]: why lazy init? is it slow?

File: selfdrive/ui/qt/widgets/cameraview.cc [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+}
+
+void CameraViewWidget::resizeGL(int w, int h) {
+  viz_rect = {bdr_s, bdr_s, w - 2 * bdr_s, h - 2 * bdr_s};
+}
+
+void CameraViewWidget::paintGL() {
+  // draw background
+
+  const QColor &color = bg_colors[ui_status];
```
Comment (adeebshihadeh) [inline selfdrive/ui/qt/widgets/cameraview.cc:158]: the widget shouldn't need a background. the frame should take up the whole widget

File: selfdrive/ui/qt/widgets/cameraview.h [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+  bool frameReceived() { return vipc_client && vipc_client->connected && last_frame != nullptr; }
+
+protected:
+  void paintGL() override;
+  void initializeGL() override;
+  void resizeGL(int w, int h) override;
+  virtual void draw() = 0;
+
+  QRect viz_rect;
+  QRect video_rect;
```
Comment (adeebshihadeh) [inline selfdrive/ui/qt/widgets/cameraview.h:30]: no viz or video rect, see above comment

File: selfdrive/ui/qt/widgets/cameraview.h [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+#include <QOpenGLWidget>
+
+#include "cereal/visionipc/visionipc_client.h"
+#include "selfdrive/common/glutil.h"
+#include "selfdrive/common/mat.h"
+#include "selfdrive/common/visionimg.h"
+#include "selfdrive/ui/ui.h"
+
+class CameraViewWidget : public QOpenGLWidget, protected QOpenGLFunctions {
+Q_OBJECT
```
Comment (adeebshihadeh) [inline selfdrive/ui/qt/widgets/cameraview.h:15]: indent

File: selfdrive/ui/qt/widgets/cameraview.cc [hunk 13]
```diff
@@ -0,0 +1,231 @@
...
+  }
+
+  // draw others
+
+  glEnable(GL_BLEND);
+  glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
+  glViewport(0, 0, vwp_w, vwp_h);
+  draw();
+  glDisable(GL_BLEND);
+  glDisable(GL_SCISSOR_TEST);
```
Comment (adeebshihadeh) [inline selfdrive/ui/qt/widgets/cameraview.cc:197]: I think we can remove this, other widgets should draw on top of it

Comment (adeebshihadeh) [reply selfdrive/ui/qt/widgets/cameraview.cc:197]: see the onroad window class for an example

File: selfdrive/ui/qt/widgets/cameraview.h [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+
+protected:
+  void paintGL() override;
+  void initializeGL() override;
+
+protected slots:
+  void updateFrame();
+
+private:
+  VisionBuf *last_frame = nullptr;
```
Comment (adeebshihadeh) [inline selfdrive/ui/qt/widgets/cameraview.h:34]: call this latest_frame

File: selfdrive/ui/qt/widgets/cameraview.h [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+#include "selfdrive/ui/ui.h"
+
+class CameraViewWidget : public QOpenGLWidget, protected QOpenGLFunctions {
+Q_OBJECT
+
+public:
+  using QOpenGLWidget::QOpenGLWidget;
+  explicit CameraViewWidget(VisionStreamType stream_type, QWidget* parent = nullptr);
+  ~CameraViewWidget();
+  bool frameReceived() { return vipc_client && vipc_client->connected && last_frame != nullptr; }
```
Comment (adeebshihadeh) [inline selfdrive/ui/qt/widgets/cameraview.h:21]: unused?

File: selfdrive/ui/qt/home.cc [hunk 16]
```diff
@@ -50,11 +50,23 @@ void HomeWindow::offroadTransition(bool offroad) {
...
+    emit previewDriverCam();
+  }
+}
+
 void HomeWindow::mousePressEvent(QMouseEvent* e) {
-  // TODO: make a nice driver view widget
-  if (QUIState::ui_state.scene.driver_view) {
-    Params().putBool("IsDriverViewEnabled", false);
-    QUIState::ui_state.scene.driver_view = false;
+  if (driver_view) {
```
Comment (adeebshihadeh) [inline selfdrive/ui/qt/home.cc:65]: don't lazy init

File: selfdrive/ui/qt/widgets/cameraview.h [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+
+signals:
+ void frameUpdated();
```
Comment (adeebshihadeh) [inline selfdrive/ui/qt/widgets/cameraview.h:24]: call this frameReceived?

File: selfdrive/ui/qt/offroad/driverview.h [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+
+#include <memory>
+
+#include <QStackedLayout>
+
+#include "selfdrive/common/util.h"
+#include "selfdrive/ui/qt/widgets/cameraview.h"
+
+class DriverViewWindow;
+class DriverViewScene : public QWidget {
```
Comment (adeebshihadeh) [inline selfdrive/ui/qt/offroad/driverview.h:11]: why two classes? I'd just combine them

File: selfdrive/ui/qt/offroad/driverview.h [hunk 19]
```diff
@@ -0,0 +1,48 @@
...
+};
+
+class DriverViewWindow : public QWidget {
+  Q_OBJECT
+
+public:
+  DriverViewWindow(QWidget *parent);
+
+signals:
+  void update(const UIState &s);
```
Comment (adeebshihadeh) [inline selfdrive/ui/qt/offroad/driverview.h:37]: let's just make a new sm for driver view to make it self contained

File: selfdrive/ui/qt/offroad/driverview.h [hunk 20]
```diff
@@ -0,0 +1,48 @@
...
+
+private:
+  Params params;
+  SubMaster sm;
+  bool frame_updated = false;
+  bool is_rhd = false;
+  QImage face_img;
+};
+
+class DriverViewWindow : public QWidget {
```
Comment (adeebshihadeh) [inline selfdrive/ui/qt/offroad/driverview.h:32]: can we combine the classes?

File: selfdrive/ui/qt/widgets/cameraview.cc [hunk 21]
```diff
@@ -0,0 +1,220 @@
...
+}
+
+void CameraViewWidget::paintGL() {
+  if (!latest_frame) {
+    glClearColor(0, 0, 0, 1.0);
+    glClear(GL_STENCIL_BUFFER_BIT | GL_COLOR_BUFFER_BIT);
+    return;
+  }
+
+  QRect video_rect = {bdr_s, bdr_s, vwp_w - 2 * bdr_s, vwp_h - 2 * bdr_s};
```
Comment (adeebshihadeh) [inline selfdrive/ui/qt/widgets/cameraview.cc:167]: the camera widget shouldn't have a border

File: selfdrive/ui/qt/offroad/driverview.cc [hunk 22]
```diff
@@ -0,0 +1,107 @@
...
+  QPainter p(this);
+  if (!frame_updated) {
+    p.setPen(QColor(0xff, 0xff, 0xff));
+    p.setRenderHint(QPainter::TextAntialiasing);
+    configFont(p, "Open Sans", 100, "Bold");
+    p.drawText(geometry(), Qt::AlignCenter, "Please wait for camera to start");
+    return;
+  }
+
+  QRect video_rect = {bdr_s, bdr_s, vwp_w - 2 * bdr_s, vwp_h - 2 * bdr_s};
```
Comment (adeebshihadeh) [inline selfdrive/ui/qt/offroad/driverview.cc:62]: just use the widget's rect

File: selfdrive/ui/qt/offroad/driverview.h (hunk 20 above)
Comment (deanlee) [reply selfdrive/ui/qt/offroad/driverview.h:32]: inherits from CameraViewWidget, or use it as a child widget of DriverViewWindow?

**Comment by adeebshihadeh** [comment]
Continued in #21063

**Pull request closed without merging**